### PR TITLE
Fix bug in 3D->2D Ellipsoid Projection

### PR DESCRIPTION
**Pull request by JiachuanXu**
This PR fix previous Issue #353 on the 3D ellipsoid projection. The projection method now works in the following way:
- Define the orientation of the galaxy eigenaxes (A-B-C) in the simulation frame, where A is the semimajor axis, and C is the semiminor axis. 
- Define the RA/Dec frame (west-north-LoS) of the galaxy in the simulation frame. 
- Transform the west-north-LoS basis vectors from the simulation frame to the galaxy body frame (A-B-C), and get the Euler angles, which rotate A-B-C into west-north-LoS
- Project the 3D ellipsoid onto the west-north plane with the Euler angles calculated in the previous step

A new degree-of-freedom (omega) is introduced in addition to (mu, phi) to describe the alignment between the 2D projected plan x-axis and the west direction. 

The new implementation has passed the existing pytest and independent code comparisons (three independent codes from Nick van Alfen, Iris Reed, and Carter Williams, agree to float32 precision with Nick's method and float64 precision with Iris and Carter's method).

 There are several points worth noting:

1. The function [mc_disk_bulge_ellipsoids](https://github.com/ArgonneCPAC/diffsky/blob/ellip_proj/diffsky/ellipsoidal_shapes/mc_disk_bulge_shapes.py#L17) now asks for the 3D comoving coordinates of galaxies. The correct coordinate to use should be the NFW repositioned coordinates, but they are not available in the current workflow in [make_lj_mock.py](https://github.com/ArgonneCPAC/diffsky/blob/ellip_proj/scripts/LJ_LC_MOCKS/make_lj_mock.py#L373). The reposition happens after adding morphology. Now the [mc_disk_bulge_ellipsoids](https://github.com/ArgonneCPAC/diffsky/blob/ellip_proj/diffsky/data_loaders/hacc_utils/lc_mock.py#L657) use the `lc_data["x/y/z"]`, which is technically inaccurate for low-z and large reposition offset satellites. 

2. Originally, a bulge-disk psi offset is manually added in `mc_disk_bulge_ellipsoids`. I don't know what's the motivation for that, but I [keep this offset in the current implementation](https://github.com/ArgonneCPAC/diffsky/blob/ellip_proj/diffsky/ellipsoidal_shapes/mc_disk_bulge_shapes.py#L169). I don't know if it's still necessary if the projection is correct. 

3. The definition of ellipticity in Diffsky is 1-q, where q=b/a is the semiminor to semimajor axes ratio. This is different from the definition of ellipticities in many weak lensing analyses (e.g., TreeCorr). This PR also includes a flag in [compute_ellipse2d](https://github.com/ArgonneCPAC/diffsky/blob/ellip_proj/diffsky/ellipsoidal_shapes/ellipse_proj_kernels.py#L93) to switch among different ellipticity definitions. 

4. This PR implements the projection by taking the full projection like sihouttee, while the previous implementation takes the z=0 intersection as the projected ellipse. Function `compute_ellipse2d` includes a flag that switches between these two kinds of projections, with the default value `envelop=True` taking the sihouttee projection. 

5. I'm not very clear on what should I do regarding fixing the [size_modeling](https://github.com/ArgonneCPAC/diffsky/tree/ca019b4bbdebc4f36eca7e74035e28630fe2bc34/diffsky/experimental/size_modeling) back to self-consistent calibration. My understanding is, the calibration is done by [disk_opt.py](https://github.com/ArgonneCPAC/diffsky/blob/ca019b4bbdebc4f36eca7e74035e28630fe2bc34/diffsky/ellipsoidal_shapes/disk_opt.py) and [bulge_opt.py](https://github.com/ArgonneCPAC/diffsky/blob/ca019b4bbdebc4f36eca7e74035e28630fe2bc34/diffsky/ellipsoidal_shapes/bulge_opt.py). They are now using the new projection `compute_ellipse2d`, so it should self-consistently calibrate the 3D size now. I don't know what else I should do.

### CODE REVIEW

**Comment by aphearin** [comment]
1. In [d0724a6](https://github.com/ArgonneCPAC/diffsky/pull/468/commits/d0724a6b47230e3baa80b557b7fa56820de17672), the function was modified to accept [mc_disk_bulge_ellipsoids](https://github.com/ArgonneCPAC/diffsky/blob/ellip_proj/diffsky/ellipsoidal_shapes/mc_disk_bulge_shapes.py#L17) `xyz_nfw` instead of `xyz`.
2. The offset between bulge-disk psi was implemented in https://github.com/ArgonneCPAC/diffsky/pull/380 so that the two orientation angles have a noisy correlation, roughly in accord with measurements made in Figure 11 of [Mandelbaum+14](https://arxiv.org/pdf/1308.4982) (see [this comment](https://github.com/ArgonneCPAC/diffsky/pull/380#issuecomment-4165744510) for details). This did not have anything to do per se with the 2d projection, and so I think we should continue to implement this noisy correlation as is for the time being.
3. The control flow implemented for the `ellipticity_type` hook may not be `jax.grad`-friendly when it comes time to re-optimize the model parameters. But we can address that in a future PR if necessary.
4. Likewise, the control flow for `envelop` may cause similar issues. Again this can be addressed in future if needed.
5. As mentioned in https://github.com/ArgonneCPAC/diffsky/issues/448, the columns `r50_disk` and `r50_bulge` are currently being used in downstream image simulation applications as the two-dimensional projected size of the disk and bulge, respectively. And so the parameters of the size--mass relation have been calibrated accordingly. As discussed with @JiachuanXu, this PR restores the original treatment of the size model such that the columns `r50_disk` and `r50_bulge` again represent the three dimensional size. In particular, on [this line](https://github.com/ArgonneCPAC/diffsky/blob/d0724a6b47230e3baa80b557b7fa56820de17672/diffsky/ellipsoidal_shapes/mc_disk_bulge_shapes.py#L119) and [this line](https://github.com/ArgonneCPAC/diffsky/blob/d0724a6b47230e3baa80b557b7fa56820de17672/diffsky/ellipsoidal_shapes/mc_disk_bulge_shapes.py#L147) of code introduced in this PR, these two columns now physically represent the 3d length of the semi-major axis. That change is expected and is what we intended. But now we will need to recalibrate [DEFAULT_DISK_SIZE_PARAMS](https://github.com/ArgonneCPAC/diffsky/blob/c0a3ef2c1fb74305e9dc4c139ade028b5c87501e/diffsky/experimental/size_modeling/smzr_disk.py#L30) and [DEFAULT_BULGE_SIZE_PARAMS](https://github.com/ArgonneCPAC/diffsky/blob/c0a3ef2c1fb74305e9dc4c139ade028b5c87501e/diffsky/experimental/size_modeling/smzr_bulge.py#L29) after another round of comparisons between image simulations of these mocks vs. LSST DP1 size measurements.

**Comment by aphearin** [comment]
At the current time, our GitHub Actions with CI workflows are not running. This is a temporary issue related to over-usage of actions by the ArgonneCPAC github organization in the month of July. I have ran pytest locally and verified that the code in [d0724a6](https://github.com/ArgonneCPAC/diffsky/pull/468/commits/d0724a6b47230e3baa80b557b7fa56820de17672) passes all unit tests.

**Review by aphearin — APPROVED** [review]
Excellent work @JiachuanXu! These changes are super clear, and so are the PR notes. My only question is whether we can simply remove the source code that begins [here](https://github.com/ArgonneCPAC/diffsky/blob/979403e4a36923b28d82460ad5f4dc70afebcab5/diffsky/ellipsoidal_shapes/ellipse_proj_kernels.py#L386) that functions like a plot-generating script, or whether we should attempt to preserve this (beyond preserving it via the git history).

**Comment by aphearin** [comment]
Closes #353

**Comment by JiachuanXu** [comment]
> Excellent work @JiachuanXu! These changes are super clear, and so are the PR notes. My only question is whether we can simply remove the source code that begins [here](https://github.com/ArgonneCPAC/diffsky/blob/979403e4a36923b28d82460ad5f4dc70afebcab5/diffsky/ellipsoidal_shapes/ellipse_proj_kernels.py#L386) that functions like a plot-generating script, or whether we should attempt to preserve this (beyond preserving it via the git history).

I agree on deleting the plotting part, that's just for sanity check and can be removed after confirming the code is working as expected.

**Review by aphearin — APPROVED** [review]

**Comment by aphearin** [comment]
Noting that in this PR we are also removing the columns `r50_disk` and `r50_bulge` from the mocks. Instead, we now ship `r50_disk_2d`, `r50_disk_3d`, `r50_bulge_2d`, `r50_bulge_3d`, to disambiguate.
